File: src/tuni_whitegoods_projector/scripts/tuni_whitegoods_projector/ui.py
import numpy as np


#class for interface image
class InterfaceUI():
    def __init__(self, ressource_id, zone, name, description, buttons): #add hidden
        self._ressource_id = ressource_id
        self._zone = zone
        self._name = name
        self._description = description
        self._list_buttons = []
        self._instructions = []
        self._hidden = False
        if len(buttons) > 0:
            self._list_buttons = buttons
        self._screen_size = (1080,1920) 
        self._interface_img = np.zeros((self._screen_size[0], self._screen_size[1], 3), np.uint8)
        self.modified_interface = True

    def add_button(self,button):
        self._list_buttons.append(button)
        self.modified_interface = True

    def add_instruction(self,inst):
        self._instructions.append(inst)
        self.modified_interface = True

    def get_zone(self):
        return self._zone

    def get_list_buttons(self):
        return self._list_buttons
    
    def get_hidden(self):
        return self._hidden

    def draw(self):
        if self.modified_interface and not self._hidden:
            for button in self._list_buttons:
                self._interface_img = button.draw_button(self._interface_img)
            for instruction in self._instructions:
                self._interface_img = instruction.draw_instruction(self._interface_img)
            self.modified_interface = False

        return self._interface_img
    
    def modify_button_color(self,button):
        for i in self._list_buttons:
            if i._id == button.id:
                i.set_button_color(button.button_color)
        self.modified_interface = True

    def get_image_interface(self):
        return self._interface_img
    
    def print_buttons(self):
        for i in self._list_buttons:
            i.print_button()
        

File: tuni_whitegoods_projector_interface/src/tuni_whitegoods_projector_interface/ui.cpp
#include "button.h"
#include "instruction.h"
#include <iostream>
#include <opencv2/opencv.hpp>
#include <vector>

class InterfaceUI {
public:
  InterfaceUI(const std::string &resource_id, const std::string &zone,
              const std::string &name, const std::string &description,
              const std::vector<Button> &buttons)
      : resource_id(resource_id), zone(zone), name(name),
        description(description), hidden(false), modified_interface(true) {
    if (!buttons.empty()) {
      list_buttons = buttons;
    }
    screen_size = cv::Size(1920, 1080);
    interface_img = cv::Mat::zeros(screen_size, CV_8UC3);
  }

  void add_button(const Button &button) {
    list_buttons.push_back(button);
    modified_interface = true;
  }

  void add_instruction(const Instruction &inst) {
    instructions.push_back(inst);
    modified_interface = true;
  }

  std::string get_zone() const { return zone; }

  const std::vector<Button> &get_list_buttons() const { return list_buttons; }

  bool get_hidden() const { return hidden; }

  cv::Mat draw() {
    if (modified_interface && !hidden) {
      for (auto &button : list_buttons) {
        button.draw_button(interface_img);
      }
      for (auto &instruction : instructions) {
        instruction.draw_instruction(interface_img);
      }
      modified_interface = false;
    }
    return interface_img;
  }

  void modify_button_color(const unity_msgs::ElementUI &button_msg) {
    for (auto &button : list_buttons) {
      if (button.get_id() == button_msg.id) {
        button.set_button_color(button_msg.button_color);
      }
    }
    modified_interface = true;
  }

  cv::Mat get_image_interface() const { return interface_img; }

  void print_buttons() const {
    for (const auto &button : list_buttons) {
      button.print_button();
    }
  }

private:
  std::string resource_id;
  std::string zone;
  std::string name;
  std::string description;
  std::vector<Button> list_buttons;
  std::vector<Instruction> instructions;
  bool hidden;
  bool modified_interface;
  cv::Size screen_size;
  cv::Mat interface_img;
};